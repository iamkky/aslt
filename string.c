#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <stdio.h>
#include <ctype.h>

#define CSIZE(class)    (sizeof(*(class)0))

typedef struct StList_struct *StList;
struct StList_struct {
	char **list;
	int size;
	StList jj;
};

typedef struct StBuffer_struct *StBuffer;
struct StBuffer_struct {
	char *buffer;
	int  allocated;
	int  size;
};

StList stListNew(int size)
{
StList self;

	if((self=malloc(CSIZE(StList)))==NULL) return NULL;

	self->size = size;
	if((self->list = malloc(sizeof(char *) * self->size)) == NULL){
		free(self);
		return NULL;
	}

	self->list[0] = NULL;

	return self;
}

int stListRegister(StList self, char *str)
{
char **tmp;
int c;

	if(self==NULL) return -1;

	for(c=0; c<=self->size; c++){
		if(self->list[c]==NULL) break;
		if(strcmp(self->list[c],str)==0) return c;
	}

	if(c + 1>=self->size){
		if((tmp = realloc(self->list, sizeof(char *) * self->size * 2))==NULL) return -1;
		self->size *= 2;
		self->list = tmp;
	}

	self->list[c] = strdup(str);
	self->list[c + 1] = NULL;
}

int stListToupper(StList self)
{
char	*p;
int	c;

	for(c=0; c<self->size; c++){
                if(self->list[c] == NULL) break;
		for(p=self->list[c]; *p; p++) *p = toupper(*p);
        }	
	return 0;
}

StBuffer stBufferNew(int alloc)
{
StBuffer self;

	if((self=malloc(CSIZE(StBuffer)))==NULL) return NULL;

	if(alloc <= 0) alloc = 128;

	self->size = 0;
	self->allocated = alloc;
	self->buffer = malloc(sizeof(char) * self->allocated);

	if(self->buffer == NULL){
		free(self);
		return NULL;
	}

	self->buffer[0] = 0;

	return self;
}

int stBufferAppend(StBuffer self, char *str)
{
char *tmp;
int len1, len2, s;

	if(self==NULL) return -1;

	len1 = strlen(self->buffer);
	len2 = strlen(str);

	if(len1 + len2 + 1 >= self->allocated){
		s = self->allocated;
		while(len1 + len2 + 1 > s) s *= 2;
		if((tmp = realloc(self->buffer, sizeof(char *) * s))==NULL) return -1;
		self->allocated = s;
		self->buffer = tmp;
	}

	strcpy(self->buffer + len1, str);
	self->size = len1 + len2;

	return len2;
}

int stBufferAppendf(StBuffer self, char *fmt, ...)
{
va_list va;
char *tmp;
int len1, len2, s;

	if(self==NULL) return -1;

	va_start(va, fmt);

	len1 = strlen(self->buffer);
	len2 = vsnprintf(self->buffer + len1, self->allocated - len1, fmt, va);

	if(len1 + len2 + 1 >= self->allocated){
		s = self->allocated;
		while(len1 + len2 + 1 > s) s *= 2;
		if((tmp = realloc(self->buffer, sizeof(char *) * s))==NULL) return -1;
		self->allocated = s;
		self->buffer = tmp;
	}

	len2 = vsnprintf(self->buffer + len1, self->allocated - len1, fmt, va);
	self->size = len1 + len2;

	return len2;
}

