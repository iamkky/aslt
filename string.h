#ifndef _RDPP_STRING_H_
#define _RDPP_STRING_H_

#define CSIZE(class)    (sizeof(*(class)0))

typedef struct StList_struct *StList;
struct StList_struct {
	char **list;
	int size;
	StList jj;
};

StList stListNew(int size);
int stListRegister(StList self, char *str);
int stListToupper(StList self);

typedef struct StBuffer_struct *StBuffer;
struct StBuffer_struct {
	char *buffer;
	int  allocated;
	int  size;
};

StBuffer stBufferNew(int alloc);
int stBufferAppend(StBuffer self, char *str);
int stBufferAppendf(StBuffer self, char *fmt, ...);

#endif
