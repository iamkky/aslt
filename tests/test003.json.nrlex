#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include "test_utils.h"

enum TOKENS {STRING=1000, INTEGER, FRACTION, EXPOENT, KEYTRUE, KEYFALSE, KEYNULL, COMMA, COLON, LBRACE, RBRACE, RBRACKET, LBRACKET};

int lex_test002(char *buffer, int *cursor, char* *value, void *extra);

int main(int argc, char **argv)
{
char *buffer;
int  size, token, cursor;
char *value;

	buffer = readToBuffer(0, 128, 1, &size);
	buffer[size]=0;
	
	cursor = 0;
	value = 0;

	while((token=lex_test002(buffer,&cursor,&value, NULL))>=0)
	{
		fprintf(stdout,"Token: %d, value=%s\n",token, value);
		if(value) free(value);
	}
}

static char *getString(char *input, int *cursor, char escape, char closure)
{
char *tmp, *newtmp, *p, last;
int  size, used, count;

	if((p = tmp = malloc(size = 128))==NULL) return NULL;
	*p = used = count = 0;

	if(*input++!=closure) return tmp;

	while((last = *p++ = *input++)){
		count++;
		if(last==closure){
			*(p-1)=0;
			break;
		}else if(last==escape && *input==closure) {
			*(p-1)=closure;
			input++;
			count++;
		}
		if(++used>=size-1) {
			if((newtmp = realloc(tmp, size = (2 * size)))==NULL) {
				free(tmp);
				return NULL;
			}
			p = (tmp = newtmp) + count;
		}
	}

	(*cursor)+=count;
	return tmp;
}


%%lex_test002 char*

%define D	0-9

true		: *((char **)value) = nrDupToken(input, size); return KEYTRUE;
false		: *((char **)value) = nrDupToken(input, size); return KEYFALSE;
null		: *((char **)value) = nrDupToken(input, size); return KEYNULL;
[		: *((char **)value) = nrDupToken(input, size); return LBRACKET;
]		: *((char **)value) = nrDupToken(input, size); return RBRACKET;
{		: *((char **)value) = nrDupToken(input, size); return LBRACE;
}		: *((char **)value) = nrDupToken(input, size); return RBRACE;
:		: *((char **)value) = nrDupToken(input, size); return COLON;
,		: *((char **)value) = nrDupToken(input, size); return COMMA;
"		: *((char **)value) = getString(input, cursor, '\\', '\"'); return STRING;
\D\D*		: *((char **)value) = nrDupToken(input, size); return INTEGER;
-\D\D*		: *((char **)value) = nrDupToken(input, size); return INTEGER;
\.\D\D*		: *((char **)value) = nrDupToken(input, size); return FRACTION;
E+\D\D*		: *((char **)value) = nrDupToken(input, size); return EXPOENT;
e+\D\D*		: *((char **)value) = nrDupToken(input, size); return EXPOENT;
E-\D\D*		: *((char **)value) = nrDupToken(input, size); return EXPOENT;
e-\D\D*		: *((char **)value) = nrDupToken(input, size); return EXPOENT;
\s		: 
\n		: 
\t		: 
\r		: 
