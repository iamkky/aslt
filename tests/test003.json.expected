Token: 1012, value=[
Token: 1009, value={
Token: 1000, value=employee
Token: 1008, value=:
Token: 1009, value={
Token: 1000, value=name
Token: 1008, value=:
Token: 1000, value=sonoo
Token: 1007, value=,
Token: 1000, value=salary
Token: 1008, value=:
Token: 1001, value=56000
Token: 1007, value=,
Token: 1000, value=married
Token: 1008, value=:
Token: 1004, value=true
Token: 1007, value=,
Token: 1000, value=description
Token: 1008, value=:
Token: 1000, value=text\twith\tescapes
Token: 1007, value=,
Token: 1000, value=extra
Token: 1008, value=:
Token: 1000, value=text with "double quotes"
Token: 1010, value=}
Token: 1007, value=,
Token: 1000, value=history
Token: 1008, value=:
Token: 1012, value=[
Token: 1000, value=2002
Token: 1007, value=,
Token: 1000, value=2003
Token: 1007, value=,
Token: 1000, value=2004
Token: 1011, value=]
Token: 1007, value=,
Token: 1000, value=values
Token: 1008, value=:
Token: 1012, value=[
Token: 1001, value=1827
Token: 1002, value=.0
Token: 1007, value=,
Token: 1001, value=1882
Token: 1002, value=.7
Token: 1007, value=,
Token: 1001, value=1889
Token: 1002, value=.5
Token: 1007, value=,
Token: 1001, value=1928
Token: 1002, value=.4
Token: 1011, value=]
Token: 1010, value=}
Token: 1007, value=,
Token: 1012, value=[
Token: 1001, value=12
Token: 1002, value=.64
Token: 1003, value=E+5
Token: 1007, value=,
Token: 1001, value=7373
Token: 1007, value=,
Token: 1001, value=-1
Token: 1002, value=.26
Token: 1003, value=e-5
Token: 1011, value=]
Token: 1011, value=]
